### PR TITLE
Ruby :  XPath Injection Query (CWE-643)

### PR DESCRIPTION
**Pull request by maikypedia**
This pull request adds a query for XPath Injection covering `rexml`, `nokogiri` and `libxml`.

Looking forward to your suggestions.

### CODE REVIEW

File: ruby/ql/lib/codeql/ruby/security/XpathInjectionQuery.qll [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+ */
+class Configuration extends TaintTracking::Configuration {
+  Configuration() { this = "Xpath Injection" }
+
+  override predicate isSource(DataFlow::Node source) { source instanceof Source }
+
+  override predicate isSink(DataFlow::Node sink) { sink instanceof Sink }
+
+  override predicate isSanitizer(DataFlow::Node node) { node instanceof Sanitizer }
+}
```
Comment (alexrford) [inline ruby/ql/lib/codeql/ruby/security/XpathInjectionQuery.qll:24]: We've recently changed how dataflow configurations are written to use a module implementing `DataFlow::ConfigSig` instead of a class extending `TaintTracking::Configuration`. There's an example of this in https://github.com/github/codeql/pull/12419/files - would you mind making this query use the new format instead, as this will save having to deprecate and port this configuration later on?

File: ruby/ql/src/experimental/xpath-injection/XpathInjection.ql [hunk 2]
```diff
@@ -0,0 +1,21 @@
+/**
+ * @name XPath query built from user-controlled sources
+ * @description Building a XPath query from user-controlled sources is vulnerable to insertion of
+ *              malicious Xpath code by the user.
```
Comment (alexrford) [inline ruby/ql/src/experimental/xpath-injection/XpathInjection.ql:4]:
```suggestion
 *              malicious XPath code by the user.
```

File: ruby/ql/lib/codeql/ruby/frameworks/Libxml.qll [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+
+    override MethodCall getACall() { result = any(LibXmlRubyXmlParserCall c).asExpr().getExpr() }
+
+    override predicate propagatesFlowExt(string input, string output, boolean preservesValue) {
+      input = "Argument[0]" and output = "ReturnValue" and preservesValue = false
+    }
+  }
+
+  /** A call that parses XML. */
+  private class LibXmlRubyXmlParserCall extends DataFlow::CallNode {
```
Comment (alexrford) [inline ruby/ql/lib/codeql/ruby/frameworks/Libxml.qll:27]: Should this extend `XmlParserCall::Range`?

File: ruby/ql/lib/codeql/ruby/frameworks/Nokogiri.qll [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+
+    override MethodCall getACall() { result = any(NokogiriXmlParserCall p).asExpr().getExpr() }
+
+    override predicate propagatesFlowExt(string input, string output, boolean preservesValue) {
+      input = "Argument[0]" and output = "ReturnValue" and preservesValue = false
+    }
+  }
+
+  /** A call that parses XML. */
+  private class NokogiriXmlParserCall extends DataFlow::CallNode {
```
Comment (alexrford) [inline ruby/ql/lib/codeql/ruby/frameworks/Nokogiri.qll:27]: Should this extend `XmlParserCall::Range`?

**Review by alexrford — CHANGES_REQUESTED** [review]
Thanks for the contribution! A few nitpicks, but generally looks good to me.

File: ruby/ql/lib/codeql/ruby/frameworks/Nokogiri.qll (hunk 4 above)
Comment (maikypedia) [reply ruby/ql/lib/codeql/ruby/frameworks/Nokogiri.qll:27]: Hi Alex! Thanks for the review, about Nokogiri I'm not sure what to do. Since this parser call is a "tiny" version of the parser call written in `XmlParsing.qll` I don't know if I can somehow import it or just copy that part to `Nokogiri.qll`

**Review by alexrford — CHANGES_REQUESTED** [review]
> Hi Alex! Thanks for the review, about Nokogiri I'm not sure what to do. Since this parser call is a "tiny" version of the parser call written in `XmlParsing.qll` I don't know if I can somehow import it or just copy that part to `Nokogiri.qll`

Ah, sorry! I hadn't noticed `XmlParsing.qll`. I see that we end up with a few nearly duplicate classes (e.g. `NokogiriXmlParserCall`, `LibXmlRubyXmlParserCall`) between `XmlParsing.qll` and the library specific files.

I'd rather avoid having these similar classes and only have one copy. Though I can see how having separate files for each library is neater - for now it seems easier to move all the XML parsing related things into `XmlParsing.qll`. The main issue with separating these out is the shared code for XML features in `XmlParsing.qll` which is `private` to avoid exposing implementation details to users of these libraries.

So in summary, my suggestion is to:
 1. Put the code from `LibXml.qll`, `Nokogiri.qll`, and `Rexml.qll` into `XmlParsing.qll`
 2. Preserve any of the existing (in `main`) implementations for `externalEntitiesEnabled()` predicates
 3. Remove the `LibXml.qll`, `Nokogiri.qll`, and `Rexml.qll` files.

This is probably the simplest way to avoid code duplication here without having to significantly refactor `XmlParsing.qll`.

**Comment by maikypedia** [comment]
Hey Alex! Sorry for the delay 😅.  About the `DataFlow::ConfigSig` I got some deprecation errors, should I change something?

**Comment by alexrford** [comment]
> Hey Alex! Sorry for the delay 😅. About the `DataFlow::ConfigSig` I got some deprecation errors, should I change something?

I think this was regarding the LDAP injection query? Should be error free now.


I'm happy to merge this PR once CI completes.

**Review by alexrford — APPROVED** [review]

**Comment by maikypedia** [comment]
> ~I think this was regarding the LDAP injection query? Should be error free now.~
> 
> I see that this was also the case for this PR - pushed a fix.
> 
> I'm happy to merge this PR once CI completes.

Thanks for the review and the merge 😄. It's always a pleasure to have you as a reviewer 🫱‍🫲 @alexrford .
